### PR TITLE
grpc stat: copy request path attributes instead of keeping them as a reference

### LINKED ISSUE #16715: Request rewriting routine corrupts memory used by grpc_stats_filter.cc
*Description*:

We use envoy to downgrade request: gRPC -> HTTP1.1
Config: https://gist.github.com/negram/1c4adae7c6cc37d8f9c02b484e09705f
When sending the following payload (a dump of gRPC request, see in "repro" steps), we found out that stats reports have corrupted data, when sending request for `com.bumble.contract.service.usersplit.UserSplitService/GetVariant`

We can see the following records (in http://localhost:9901/stats/prometheus)
```
# TYPE envoy_cluster_grpc_bumble_contract_service_u_ersplit_Us_request_message_count counter
envoy_cluster_grpc_bumble_contract_service_u_ersplit_Us_request_message_count{envoy_grpc_bridge_method="phtml?package=com",envoy_grpc_bridge_service="grpcserver",envoy_cluster_name="service_app_http"} 1
```
**TYPE** and **envoy_grpc_bridge_method** look not  perfectly right here.

Further investigation showed that grpc_stats_filter.cc uses `absl::string_view` to hold service/method, basically it stores info where this data is held in memory. But routing procedure does in-place rewriting and breaks the data used by the stat module.

*Repro steps*:

We can use this payload to check how it works
```
echo "UFJJICogSFRUUC8yLjANCg0KU00NCg0KAAAkBAAAAAAAAAIAAAAAAAMAAAAAAAQAQAAAAAUAQAAAAAYAACAA/gMAAAABAAAECAAAAAAAAD8AAQAACAYAAAAAAAAAAAAAAAAAAAEfAQQAAAABhoNACjphdXRob3JpdHkPbG9jYWxob3N0OjUxMDUxQAU6cGF0aEIvY29tLmJ1bWJsZS5jb250cmFjdC5zZXJ2aWNlLnVzZXJzcGxpdC5Vc2VyU3BsaXRTZXJ2aWNlL0dldFZhcmlhbnRAAnRlCHRyYWlsZXJzQAxjb250ZW50LXR5cGUQYXBwbGljYXRpb24vZ3JwY0AKdXNlci1hZ2VudDNncnBjLW5vZGUvMS4yNC4yIGdycGMtYy84LjAuMCAob3N4OyBjaHR0cDI7IGdhbmdlcylAFGdycGMtYWNjZXB0LWVuY29kaW5nFWlkZW50aXR5LGRlZmxhdGUsZ3ppcEAPYWNjZXB0LWVuY29kaW5nDWlkZW50aXR5LGd6aXAAAAQIAAAAAAEAAAAFAAAiAAEAAAABAAAAAB0KAghkEgcKBUhlbGxvGg4KBUhlbGxvEgVIZWxsbwAABAgAAAAAAAAAAAU=" > grpc-usersplit.dump
```

I used the last master build to reproduce the issue, 

```
git checkout aede3f4a736a1364b3ea18d8286379e48bbc45f9
```

Build on Mac (on Linux we see the same behaviour)
```
time bazel build --jobs 20 --local_cpu_resources=36 --local_ram_resources=HOST_RAM -c dbg --spawn_strategy=standalone --strategy=BuildRunner=local //source/exe:envoy-static  --strip="never"
```

Run under debugger (config was given above [php-grpc-gw-envoy.yaml](https://gist.github.com/negram/1c4adae7c6cc37d8f9c02b484e09705f))
```
lldb -- bazel-bin/source/exe/envoy-static -c php-grpc-gw-envoy.yaml
```

Set a breakpoint on the filter, in the place where the service/method has just appeared and run
```
(lldb) b source/extensions/filters/http/grpc_stats/grpc_stats_filter.cc:157
(lldb) r
```

Send the request to Envoy (from another console)
```
$ nc 127.0.0.1 51051 < grpc-usersplit.dump
```

And we can see the perfectly correct service/method at this point
```
Process 72537 stopped
* thread #23, stop reason = breakpoint 1.1
    frame #0: 0x00000001019e45bf envoy-static`Envoy::Extensions::HttpFilters::GrpcStats::(anonymous namespace)::GrpcStatsFilter::decodeHeaders(this=0x00000000106308d8, headers=0x0000000010678000, (null)=false) at grpc_stats_filter.cc:157:31
...
(lldb) p request_names_->method_
(Envoy::Stats::Element) $0 =  Active Type = Envoy::Stats::DynamicName  {
  Value = {
    absl::string_view = (__data = "GetVariant\x9a\U00000013\U00000001", __size = 10)
  }
}

(lldb) bt
* thread #23, stop reason = breakpoint 1.1
  * frame #0: 0x00000001019e45bf envoy-static`Envoy::Extensions::HttpFilters::GrpcStats::(anonymous namespace)::GrpcStatsFilter::decodeHeaders(this=0x00000000106308d8, headers=0x0000000010678000, (null)=false) at grpc_stats_filter.cc:157:31
    frame #1: 0x000000010bc598c5 envoy-static`Envoy::Http::ActiveStreamDecoderFilter::decodeHeaders(this=0x00000000101b5880, headers=0x0000000010678000, end_stream=false) at filter_manager.h:286:43
    frame #2: 0x000000010bc4ba5a envoy-static`Envoy::Http::FilterManager::decodeHeaders(this=0x00000000100c6478, filter=0x0000000000000000, headers=0x0000000010678000, end_stream=false) at filter_manager.cc:527:44
    frame #3: 0x000000010bb81146 envoy-static`Envoy::Http::FilterManager::decodeHeaders(this=0x00000000100c6478, headers=0x0000000010678000, end_stream=false) at filter_manager.h:784:5
    frame #4: 0x000000010bb7f9e2 envoy-static`Envoy::Http::ConnectionManagerImpl::ActiveStream::decodeHeaders(this=0x00000000100c6400, headers=nullptr, end_stream=false) at conn_manager_impl.cc:1074:19
    frame #5: 0x000000010bec72ef envoy-static`Envoy::Http::Http2::ConnectionImpl::ServerStreamImpl::decodeHeaders(this=0x0000000010674000) at codec_impl.cc:356:21
    frame #6: 0x000000010bed3083 envoy-static`Envoy::Http::Http2::ConnectionImpl::onFrameReceived(this=0x0000000002e82ea0, frame=0x00000000106601a0) at codec_impl.cc:832:15
    frame #7: 0x000000010bee7930 envoy-static`Envoy::Http::Http2::ConnectionImpl::Http2Callbacks::Http2Callbacks(this=0x0000000010660000, (null)=0x0000000010660000, frame=0x00000000106601a0, user_data=0x0000000002e82ea0)::$_17::operator()(nghttp2_session*, nghttp2_frame const*, void*) const at codec_impl.cc:1304:64
    frame #8: 0x000000010bee78f5 envoy-static`Envoy::Http::Http2::ConnectionImpl::Http2Callbacks::Http2Callbacks((null)=0x0000000010660000, frame=0x00000000106601a0, user_data=0x0000000002e82ea0)::$_17::__invoke(nghttp2_session*, nghttp2_frame const*, void*) at codec_impl.cc:1303:19
```

Now we can set a watchpoint to catch the moment where it gets broken, and continue the execution:
```
(lldb) watchpoint set expression  ((absl::string_view*)&request_names_->method_)->__data

(lldb) (lldb) c
Process 72537 resuming

Watchpoint 1 hit:
old value: 0x6169726156746547
new value: 0x6169726156746565
Process 72537 stopped
* thread #23, stop reason = watchpoint 1
    frame #0: 0x000000010a5366db envoy-static`absl::inlined_vector_internal::IteratorValueAdapter<std::__1::allocator<char>, char const*>::AssignNext(this=0x000070000ad078f8, assign_at="eetVariant\x9a\U00000013\U00000001") at inlined_vector.h:127:5
Target 0: (envoy-static) stopped.
```

Now we can see that "**G**etVariant" turned into "**e**etVariant" and it will be changed further.

Backtrace:
```
(lldb) bt

  * frame #0: 0x000000010a5366db envoy-static`absl::inlined_vector_internal::IteratorValueAdapter<std::__1::allocator<char>, char const*>::AssignNext(this=0x000070000ad078f8, assign_at="eetVariant\x9a\U00000013\U00000001") at inlined_vector.h:127:5
    frame #1: 0x000000010a5361fe envoy-static`void absl::inlined_vector_internal::AssignElements<char*, absl::inlined_vector_internal::IteratorValueAdapter<std::__1::allocator<char>, char const*>, unsigned long>(assign_first="/grpcserver.phtml?package=com.bumble.contract.service.useetVariant\x9a\U00000013\U00000001", values_ptr=0x000070000ad078f8, assign_size=66) at inlined_vector.h:97:17
    frame #2: 0x0000000110923015 envoy-static`void absl::inlined_vector_internal::Storage<char, 128ul, std::__1::allocator<char> >::Assign<absl::inlined_vector_internal::IteratorValueAdapter<std::__1::allocator<char>, char const*> >(this=0x0000000010396fa8, values=(it_ = "ersplit.UserSplitService&method=GetVariant"), new_size=98) at inlined_vector.h:526:3
    frame #3: 0x00000001108e9019 envoy-static`void absl::InlinedVector<char, 128ul, std::__1::allocator<char> >::assign<char const*, (this=0x0000000010396fa8, first="/grpcserver.phtml?package=com.bumble.contract.service.usersplit.UserSplitService&method=GetVariant", last="")0>(char const*, char const*) at inlined_vector.h:514:14
    frame #4: 0x00000001108e9c1f envoy-static`Envoy::Http::HeaderString::setCopy(this=0x0000000010396fa8, data="/grpcserver.phtml?package=com.bumble.contract.service.usersplit.UserSplitService&method=GetVariant", size=98) at header_map_impl.cc:125:21
    frame #5: 0x00000001108ea074 envoy-static`Envoy::Http::HeaderString::setCopy(this=0x0000000010396fa8, view=(__data = "/grpcserver.phtml?package=com.bumble.contract.service.usersplit.UserSplitService&method=GetVariant", __size = 98)) at header_map_impl.cc:130:9
    frame #6: 0x00000001009ae68c envoy-static`Envoy::Http::TypedHeaderMapImpl<Envoy::Http::RequestHeaderMap>::setInline(this=0x0000000010678000, handle=Envoy::Http::TypedHeaderMapImpl<Envoy::Http::RequestHeaderMap>::Handle @ 0x000070000ad07de8, value=(__data = "/grpcserver.phtml?package=com.bumble.contract.service.usersplit.UserSplitService&method=GetVariant", __size = 98))0>, std::__1::basic_string_view<char, std::__1::char_traits<char> >) at header_map_impl.h:439:19
    frame #7: 0x00000001009b1fc1 envoy-static`Envoy::Http::RequestHeaderMapImpl::setPath(this=0x0000000010678000, value=(__data = "/grpcserver.phtml?package=com.bumble.contract.service.usersplit.UserSplitService&method=GetVariant", __size = 98)) at header_map_impl.h:478:3
    frame #8: 0x000000010c934252 envoy-static`Envoy::Router::RouteEntryImplBase::finalizePathHeader(this=0x00000000100fc000, headers=0x0000000010678000, matched_path=(__data = "/", __size = 1), insert_envoy_original_path=true) const at config_impl.cc:708:11
    frame #9: 0x000000010c9394e0 envoy-static`Envoy::Router::PrefixRouteEntryImpl::rewritePathHeader(this=0x00000000100fc000, headers=0x0000000010678000, insert_envoy_original_path=true) const at config_impl.cc:1079:3
    frame #10: 0x000000010c933041 envoy-static`Envoy::Router::RouteEntryImplBase::finalizeRequestHeaders(this=0x00000000100fc000, headers=0x0000000010678000, stream_info=0x00000000100c6568, insert_envoy_original_path=true) const at config_impl.cc:608:5

```

Finally, when we record the stat, i.e. in `GrpcStatsFilter::encodeTrailers` the data is not something we wanted to have. 
I was able to crutch-around it with a patch: [copy-stat-data.patch](https://gist.github.com/negram/cff7822f8726c685f3972b2ad8cb66e1) - basically, I just copy string once the method/service is decoded and store it in the string-views

It's pretty far from perfect but it seems it does the trick. I believe, there must be a better solution

### PR DESCRIPTION
**Pull request by negram**
Fixes #16715

The service/method attributes might be changed due to the routing config.

It looks safer to preserve them to be able able to get the
correct statistics once the request is over.

Signed-off-by: Mikhail Galanin <mikhail.galanin@team.bumble.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by negram** [comment]
Basically, this fix has been working on our prod for 4 weeks.
It seems to be less efficient than the initial implementation but feels more reliable - contains fewer potential corner cases.
The only thing I'm not sure about is tests - should we provide more complex tests or this looks fine?

File: source/common/grpc/context_impl.cc [hunk 1]
```diff
@@ -85,12 +85,13 @@ void ContextImpl::chargeUpstreamStat(const Upstream::ClusterInfo& cluster,
...
   absl::optional<Common::RequestNames> request_names = Common::resolveServiceAndMethod(path);
+
   if (!request_names) {
     return {};
   }
 
-  Stats::Element service = Stats::DynamicName(request_names->service_);
-  Stats::Element method = Stats::DynamicName(request_names->method_);
+  Stats::Element service = Stats::DynamicSavedName(request_names->service_);
+  Stats::Element method = Stats::DynamicSavedName(request_names->method_);
```
Comment (rojkov) [inline source/common/grpc/context_impl.cc:94]: Nit: I think it's worth adding a comment here that these new stat names may outlive the values in `request_names_` thus need to be turned into copies.

File: source/common/stats/utility.cc [hunk 2]
```diff
@@ -52,6 +52,7 @@ struct ElementVisitor {
   // Overloads provides for absl::visit to call.
   void operator()(StatName stat_name) { stat_names_.push_back(stat_name); }
   void operator()(absl::string_view name) { stat_names_.push_back(pool_.add(name)); }
+  void operator()(const std::string& name) { stat_names_.push_back(pool_.add(name)); }
```
Comment (rojkov) [inline source/common/stats/utility.cc:55]: Not sure this overloading is needed.

File: source/common/stats/utility.h [hunk 3]
```diff
@@ -30,6 +30,25 @@ class DynamicName : public absl::string_view {
...
+ *
+ * Specifically, this class should be used only when the content of the string
+ * can be changed between the object creation and usage.
+ */
+class DynamicSavedName : public std::string {
+public:
+  // This is intentionally left as an implicit conversion from string_view to
+  // make call-sites easier to read, e.g.
+  //    Utility::counterFromElements(*scope, {DynamicName("a"), DynamicName("b")});
+  explicit DynamicSavedName(const std::string& str) : std::string(str) {}
```
Comment (rojkov) [inline source/common/stats/utility.h:48]: Can we drop this constructor?

File: test/common/grpc/context_impl_test.cc [hunk 4]
```diff
@@ -73,8 +74,8 @@ TEST(GrpcContextTest, ResolveServiceAndMethod) {
   absl::optional<Context::RequestStatNames> request_names =
       context.resolveDynamicServiceAndMethod(path);
   EXPECT_TRUE(request_names);
-  EXPECT_EQ("service_name", absl::get<Stats::DynamicName>(request_names->service_));
-  EXPECT_EQ("method_name", absl::get<Stats::DynamicName>(request_names->method_));
+  EXPECT_EQ("service_name", absl::get<Stats::DynamicSavedName>(request_names->service_));
+  EXPECT_EQ("method_name", absl::get<Stats::DynamicSavedName>(request_names->method_));
```
Comment (rojkov) [inline test/common/grpc/context_impl_test.cc:78]: This test case shouldn't fail with `Stats::DynamicName`. Does it?

File: test/common/grpc/context_impl_test.cc [hunk 5]
```diff
@@ -87,5 +88,28 @@ TEST(GrpcContextTest, ResolveServiceAndMethod) {
   EXPECT_FALSE(context.resolveDynamicServiceAndMethod(path));
 }
 
+TEST(GrpcContextTest, ResolveServiceAndMethodNotChanged) {
```
Comment (rojkov) [inline test/common/grpc/context_impl_test.cc:91]: Nit: The name of the test case feels a bit misleading as the names of the request are actually changed.

**Review by rojkov — COMMENTED** [review]
Thanks! This looks reasonable. Could you please also extend `test/common/stats/utility_test.cc` to cover the new type?

File: test/common/grpc/context_impl_test.cc (hunk 4 above)
Comment (negram) [reply test/common/grpc/context_impl_test.cc:78]: **service_**/**method_** is _absl::variant_ now storing  `DynamicSavedName`, if we try extracting `DynamicName` it's gonna produce `C++ exception with description "Bad variant access"`

File: source/common/stats/utility.h (hunk 3 above)
Comment (negram) [reply source/common/stats/utility.h:48]: it seems like we can, it works even better, thanks

**Comment by negram** [comment]
> Could you please also extend `test/common/stats/utility_test.cc` to cover the new type?

I'm probably gonna need help with this. Is what I did enough? Maybe it might be done somehow better?

File: test/common/grpc/context_impl_test.cc (hunk 5 above)
Comment (negram) [reply test/common/grpc/context_impl_test.cc]: Does this become clearer now?

File: source/common/stats/utility.cc (hunk 2 above)
Comment (negram) [reply source/common/stats/utility.cc]: Indeed, removed

File: source/common/grpc/context_impl.cc (hunk 1 above)
Comment (negram) [reply source/common/grpc/context_impl.cc:97]: Added, would take a look, please?

**Comment by rojkov** [comment]
> I'm probably gonna need help with this. Is what I did enough? Maybe it might be done somehow better?

This is probably good now. I just wanted to make sure that the unit test fully covers the unit under test.

**Review by rojkov — APPROVED** [review]
Thanks! Looks good to me.

/assign-from @envoyproxy/envoy-maintainers
for a second pass.

**Comment by rojkov** [comment]
/assign-from @envoyproxy/senior-maintainers 
for a second pass.

**Comment by alyssawilk** [comment]
Oops sorry for the delay - Matt's out.  @lizan can you do a pass please?

File: source/common/stats/utility.h [hunk 6]
```diff
@@ -38,7 +56,7 @@ class DynamicName : public absl::string_view {
  * hide the memory management of the joined storage, and they allow easier
  * co-mingling of symbolic and dynamic stat-name components.
  */
-using Element = absl::variant<StatName, DynamicName>;
+using Element = absl::variant<StatName, DynamicName, DynamicSavedName>;
```
Comment (lizan) [inline source/common/stats/utility.h:59]: This is not allowed:

https://en.cppreference.com/w/cpp/utility/variant
> As with unions, if a variant holds a value of some object type T, the object representation of T is allocated directly within the object representation of the variant itself. Variant is not allowed to allocate additional (dynamic) memory.

while `std::string` allocates additional memory to hold the string thus it may be leaked.

**Review by lizan — CHANGES_REQUESTED** [review]

File: source/common/stats/utility.h (hunk 6 above)
Comment (rojkov) [reply source/common/stats/utility.h:59]: I interpret it as "like unions a variant allocates enough memory to contain a value of the biggest type T thus doesn't need to allocate more (dynamically) when a value of a smaller type gets replaced with another value of a bigger type".
As with sts::string a variant needs to store its control structure only which doesn't change. And unlike unions variants call the destructors of Ts. So, no leaks are expected.

Comment (lizan) [reply source/common/stats/utility.h:59]: Ah that makes more sense, my bad.

**Review by lizan — APPROVED** [review]
